File: Algorithms/C++/86_Partition_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode less_dummy(0);
        ListNode other_dummy(0);

        ListNode* less = &less_dummy;
        ListNode* other = &other_dummy;

        while (head != NULL) {
            if (head->val < x) {
                less->next = head;
                less = less->next;
            } else {
                other->next = head;
                other = other->next;
            }
            head = head->next;
        }

        other->next = NULL;
        less->next = other_dummy.next;

        return less_dummy.next;
    }
};

File: Algorithms/Swift/Linked_List/86_Partition_List.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func partition(head: ListNode?, _ x: Int) -> ListNode? {
        if head == nil {
            return head
        }

        let lessDummy = ListNode(0)
        let greaterEqualDummy = ListNode(0)
        var lessNode: ListNode? = lessDummy
        var greaterEqualNode: ListNode? = greaterEqualDummy

        var node = head
        while node != nil {
            if node!.val < x {
                lessNode!.next = node
                lessNode = lessNode!.next
            } else {
                greaterEqualNode!.next = node
                greaterEqualNode = greaterEqualNode!.next
            }
            node = node!.next
        }

        lessNode!.next = greaterEqualDummy.next
        greaterEqualNode!.next = nil

        return lessDummy.next
    }
}
